### ISSUE TITLE
almost weekly useful materials - 2025/1/15 -

### ISSUE THREAD

**Issue by GENZITSU** [GENZITSU/UsefulMaterials#157]

**Comment by GENZITSU** [comment]
[Amazon Bedrock Knowledge Bases の新機能で PDF を読む](https://qiita.com/cyberBOSE/items/d9d169f376f5fe13ada4)

ファンデーションモデルパーサーを使うと日本語のドキュメントかつ絵付きのページもmarkdownに落としてくれる模様

**Comment by GENZITSU** [comment]
[Qwen2-VLの画像エンコーダーを使って日本語CLIPモデルを学習する｜oshizo ](https://t.co/ufmojSQGvX)
取り組みとして面白かったのでメモだけ

まだちゃんと理解できてない

**Comment by GENZITSU** [comment]
[コンサルが「最初の3年間」で学ぶコト　知らないと一生後悔する99のスキルと5の挑戦](https://www.socym.co.jp/book/1398)

ながいので章ごとに分割、第一章。

## VS思考

絶妙に迷う選択肢をちゃんと考え込んだ上で、取捨選択することで思考が深まり、選択肢の質が上がり、選んだ選択を実行するモチベーションも上がる

## 辻褄思考

ものごとの背景を理解しようとする思考(と言うか姿勢)が大事

## 相談した人には、必ず結果を報告する

結果を報告することで相談相手との関係が深まる。
誰か紹介者が挟まるときは、その紹介者にも報告

## 質問に答える時はまず論点に答える

論点に答えているかのチェックは、答えから質問に帰れるか
色々補足したくなることが多いが、まずはシンプルに答えてから。

## 構造を持って話す

箱を作ってから、中に代入していくような話し方、説明の仕方。
箱の個数に加えて、カテゴリー分け, それを見せる順番まで考慮できているとグッド。
ただし、時と場合によっては(特に相手の方が優秀な時は気にせずどんどんしゃべった方が良い時もある)
ドキュメンテーションにおいても大事そう

## アウトプットを作成するための基本プロセス

論点 → サブ論点 → タスク → スケジュール → 作業 → アウトプット

論点, サブ論点設計に頭をちゃんと使う

サブ論点というのは解きたい課題に対して、出てくる疑問点のようなもの。まだ判明していない事柄など。

逆にいえば、各タスクが何を明らかにするためのものかを見失わないように

## メリデメ / プロコンで丸めすぎない

何かを取捨選択するときの評価基準をぼやかすことになる (と言うよりまだ曖昧な状態をスルーしてしまうことに繋がる)ので評価基準の軸を括り出すことを意識する

## インプットが成否を分ける

garbage in garbage outとうより、良いアウトプットを出すためには良い材料がいるよということ。
その材料調達部分に優位性が生まれる。

## 議事録の段階的進化

- 発言の構造化
- ネクストステップのくくりだし
- その場の空気感の追加
- 論点ベースでの整理
- 論点の変化 / 仮説の進化に基づいた整理
- 新たな論点の整理

## +2度

+2度のテンションで取り組んでいた方が絡みやすい

## 怒られた時こそ接近する

- 怒られたから / 注意されたから と言っていじけない。そういう時こそ食いしばる、接近する。
- 詰められた時こそ、XXにまた時間ください

## 誤字脱字

- 誤字脱字があるとその資料の信用度は半減する
- 最良の誤字脱字チェックは音読

## 書いてから削る

まず制限文字数などを気にせず書き出す。そこから削りながら思考を深める。

## 考えると描くを明確に分ける。

具体的なタスクに取り掛かる前に、アウトプット作る前に、考える時間をちゃんと取ろう。

## wordでスライド構成を考え切ってからpptに落とす

上のやつと似てる

**Comment by GENZITSU** [comment]
[コンサルが「最初の3年間」で学ぶコト　知らないと一生後悔する99のスキルと5の挑戦](https://www.socym.co.jp/book/1398)

ながいので章ごとに分割、第二章。

## How のインサイト

- 何かのタスクを行う際に、普通ならやらないところまでできると付加価値がでる
- 泥臭いこととか、負荷が高いこと、手順な面倒なこと、など
- google検索は検索ワードに関係ないものがで始める20pあたりまで見る気概で

## 闇雲に手を動かすのは論外

- 時間に追われると、手を動かすことを優先しがち
- 手を動かす前に、なんのために、どういう順序で、どういう終了条件かを明確にする

## 活字に落とす重要性

議事録(というより、論点などを整理するところまで消化した議事メモ)はとても価値が高い
思考は文字に落とすことで伝染力を持つ

## １次情報にあたる

誰かに噛み砕かれた情報には何かしらのバイアスがのっている

## 1 + 1 + 1 = 3

同じテーマに書かれた記事でも、3箇所のソースを見ることでより深く多面的にものごとを知ることができたりする

## エグゼクティブサマリーに命をかける

まず何を伝えたいのかから固める
思考を言語を練りに練ったものがエグゼクティブサマリーとなる

## 一点豪華主義

全てが平均ちょい上よりも、何か一つに狂気を込めた方が面白いものが出来上がる

## スタンスを取る

ケースバイケースというのは素人でも言えるので、わからないときでもあえてスタンスをとった方が良い
(むろん嘘はダメ)

## 田の字

マトリックスともいう。 A vs Bを2軸にしたもの

## 何をレビューしてもらうべきか

- 自分の仕事を減らすためのレビューでは意味がない
- 上司がやるであろうステップに沿った要所要所のレビューであれば意味が出てくる

## タスクの論点に割戻す

- そのタスクをなぜやるのか、何を明らかにしたいのかを常に考える癖をつける
- 論点に割戻したあと、サブ論点
- タスクの完了条件が変わったり、より良いやることが見つかるかもしれない

## もうイーハンつける

頼まれた仕事の120点を取る + 必要そうなことを自発的に追加する

## too all ではなく too sameone

- 誰に何を届けるためのアウトプットかを意識する
- 参加者が異なれば、興味関心、論点、評価基準も異なる。
- それにあった伝え方、準備が必要になる

## ロジックではなくストーリー

SCや意思決定者向けの説明はストーリーで語っていく必要がある

## 常にベストエフォート

- 考え切れてないですが,,, 粗々ですが,,, などの前置きで期待値を下げて自己防衛しても何も進まないし、格が落ちるだけ
- 常に、その時点での最高傑作ですという気概で

(WIP)

**Comment by GENZITSU** [comment]
正しい自動化のステップ

本当にelon muskが言ったのかはよくわからないが、なるほど、と思ったので。

- まず、やるべきこと、要求事項を最小限にする
- 今のプロセスで不要なものを全て取り除く
- 必要なプロセスだけになった上で、それを単純化する
- プロセスを高速化させる (←よくわからない)、ただし上記3つをやってから
- ここまでやって自動化を考える。ただ自動化に急いてはいけない。本当に自動化するべきものだけを見定める。


やるべきこと、最適化するべきことを見定めないと、意味ないことに労力を投下することになるよねということかな


こちらがソースっぽい 確かに言ってる。: https://x.com/w1991e/status/1867777999145513021

元の動画は2021年ものらしい: https://youtu.be/t705r8ICkRw?si=r_beHdPEhv0mgObO

**Comment by GENZITSU** [comment]
[関数とかクラスとかを切り出すときに考えていること](https://zenn.dev/kohii/articles/2c1126389d914a)

おもろいと思ったところだけ抜粋

### コードのデザインと UI のデザインはだいたい同じ

UI デザインでは、ユーザーが知らなくてもいい技術や複雑さを隠蔽し、よく練られた別の概念として見せ、直感的に理解・操作できるようにします。
関数、クラス、コンポーネントやモジュールをデザインするのも、「どういう詳細をどういうインターフェイスの裏に押し込めるか」を考えるという点で UI デザインと同じです。

### 実装を読みたいと思わせたら負け

よくできたAPIやライブラリは、中の実装を読まなくてもインターフェイス（I/F）から何をしているかやどのように使えばいいかがわかります。

コードを切り出し、インターフェイスの裏側に押し込めるのは、「ここから先は読まなくていい」「ここから先のことは考えなくていい」部分を作ることが関数化

実装を読みに行かないと理解ができないのであれば、極端に言えばその切り出しは失敗しています。

### その関数やクラス自身の都合でインターフェイスを決める

- 関数やクラスが「呼び出し元の都合も考慮しなければならない」と単一責任の原則を実現しにくくなる
- 「自身の責務を全う」することに注力すべきで、呼び出す側の都合ではなく、呼び出される関数自身が必要なパラメータや戻り値を規定するのが良い

### 巻き込む文脈の量を極力減らす

↑と同じだが、関数を適切に実行するために必要な文脈を極力削ぎ落とす

ここでいう文脈とは
> - 実行環境やインフラ、フレームワークの文脈
> - 呼び出し元、呼び出し元の呼び出し元、…の文脈
> - 特定の業務フローやユースケースの文脈
> etc…

### 切り出しのオーバーヘッドを気にする

切り出しのコストも考えて、本当に関数を作るべきかを一向する

### コメントはやや親切目に書く

利点としては
> - 人によって「良い設計」の方向性や到達点が異なる
> - 読み解く時間の短縮
>     - 良い設計は確かに意味を推測しやすいが、それでも労力がかかる
> - Why の記録
>     - 諸事情で不完全またはトリッキーなロジックを書かざるをえないときがある
>     - 他人や未来の自分へ「なぜ今こういう設計判断をしたのか」を伝えるためにコメントを書く
> - コメントを書くことで、客観的なわかりにくさに気づく

### Think Twice

関数やクラスのインターフェイスが妥当かどうかは、「関数やクラス側の視点」と「それを使う側の視点」の両方を行き来して考えるとよい

### 力を入れるところと手を抜くところ

全ての関数に力を入れてもしょうがないので、たくさん利用するもの、長生きしそうなもの、基幹となってるものに注力

**Comment by GENZITSU** [comment]
[Gemini2.0とStreamlitでお手軽なグラウンディング搭載の生成アプリ作成](https://note.com/masa_kazama/n/n8eff1a4be1da)

サンプルコードもありがたいが、個人的にはgemini apiで検索もよしなにやってくれてることに驚いた


```python
# Google検索を利用
google_search_tool = types.Tool(google_search = types.GoogleSearch())

if generate_button:
    response = client.models.generate_content(
        model='gemini-2.0-flash-exp', contents=user_prompt,
        config=types.GenerateContentConfig(
            tools=[google_search_tool]
        ),
    )
    st.write(response.text)
```

**Comment by GENZITSU** [comment]
[awesome-chatgpt-prompts](https://github.com/f/awesome-chatgpt-prompts)

インフルエンサー(?)が共有してるchatgptにOOをしてもらうときのプロンプト集。

![スクリーンショット 2025-01-11 10 52 21](https://github.com/user-attachments/assets/c0f60069-d573-48b4-a42f-fe58f0b2d50e)

**Comment by GENZITSU** [comment]
[2024年にエンジニアリンググループミーティングで紹介した書籍の紹介](https://www.m3tech.blog/entry/engineering-books-2024)

いつか読むかもということでメモ

特に気になったものだけ、こちらに抜粋

### 熟達論 - 人はいつまでも学び、成長できる

為末大氏の書籍です。スポーツにおいて「真のトップランナー」へ成長する過程を考察しています。エンジニアのキャリアに通づる所も多いですし、何かを突き詰める全ての成長論に通づる良書です

### 1兆ドルコーチ

最後は、心にガツンと来る一冊です。あなたがやらなければ何も変わらない、ただそれだけの事を丁寧に伝えてくれます。私も年末読み返したい至極の一冊です。

**Comment by GENZITSU** [comment]
[Kaggle画像コンペでやっていること③](https://qiita.com/Kmat67916008/items/6db4446698a58a6b3f0d)

loss, augmentationの話

## loss

### 物理現象を意識させるロス

> RGB画像から奥行きを予測するタスクではDepth Smoothness Lossと呼ばれるLossがよく活用される
>『奥行きの変化が起きるのは色味の変化が大きいところ』だという常識をモデルに意識させるためのlossで、入力したRGB画像と予測した奥行き画像との画素間の変化量を用いて、RGBの変化量が少ない箇所での奥行き変化に対して軽いペナルティを与えます。

> 保存則の利用
> 地球への流入する熱量と流出する熱量の差分は、地球のどこかに蓄えられているので、それらをLossとして定義することができます。たとえばloss = (流入 - 流出 - 地球蓄え)**2のようにペナルティを与えられます

### サブタスクとしての補助ロス

> 分類や回帰などの本来解きたいタスクと並列に、そのドメインにとって重要な情報をサブタスク的に学習させることで、メインのタスクに対するパフォーマンスを高めることができる

> [kaggleでおこなわれたクジラの識別コンペ](https://www.kaggle.com/competitions/happy-whale-and-dolphin/overview)では、個体識別IDとともに、species(種族)が与えられていました。コンペの目的は個体分類ですので、種族を分類すること自体は求められていませんが、これをサブタスクとして解かせることで精度を上げたチームが多かった

> また、[チューリッヒ工科大の論文(CoRL2024)](https://arxiv.org/abs/2409.04882)では、扉を開けるタスクの学習において、ロボットの動きとは直接関係のない扉に関する情報を予測させ、それに対してEstimation Lossとして学習させています。
直接最終的な予測をすることが難しいようなタスクで、その判断の助けになるようなものを補助的に学習するときは効きやすいですね。

### 予測対象よりも難易度の高い教師データを活用する

分類問題だが、segmnetationマスクが与えられている例

> 逆に、ラベルよりも情報が少ないタスクを解かせても補助ロスは効きにくいことも多いです。極端な例ですが、犬の画像のセグメンテーションをするときに、画像に犬が含まれるかどうかのサブタスクを解いてもそれほど効果はありません。

### ラベルからサブタスクを作る

> 先日の[第18回atmaCup](https://www.guruguru.science/competitions/25/)では自車の未来の位置推定問題でしたが、位置だけでなく速度や加速度をサブタスクで学習する人がいました。たしかに、『青信号なら加速』『前に車両がいる場合には避ける』など、速度や加速度を用いることで学習しやすくなる要素がありそうなことが想像できます。

### 補助ロスの注意点

メインタスクに悪影響を与える時もある
対策としては
- 重みの調整
- チャネル数を増やす (←よくわからない)
- サブタスクとメインタスクでブランチを分ける ( 特徴抽出後のFFNかませる的な)
- 出力の制約として組み込む
- 諦める

## augmentation

- augmentationによってcontextやラベルが変わらないかを考える
- よく起きる状況を再現する
- 結局は全部試すことになる

## 感想

補助ロスにまだ慣れ親しんでいないので、ものにしたい。テーブルデータやNLPでもうまくできんかしら
